### PR TITLE
[RISCV][MC] Support experimental extensions Zvbc32e and Zvkgs

### PR DESCRIPTION
**Pull request by wangpc-pp**
These two extensions add addtional instructions for carryless
multiplication with 32-bits elements and Vector-Scalar GCM
instructions.

Please see https://github.com/riscv/riscv-isa-manual/pull/1306.

### CODE REVIEW

File: llvm/docs/RISCVUsage.rst [hunk 1]
```diff
@@ -302,6 +302,9 @@ The primary goal of experimental support is to assist in the process of ratifica
 ``experimental-zicfilp``, ``experimental-zicfiss``
   LLVM implements the `1.0 release specification <https://github.com/riscv/riscv-cfi/releases/tag/v1.0>`__.
 
+``experimental-zvbc32e``, ``experimental-zvkgs``
+  LLVM implements the `0.7 release specification <https://github.com/user-attachments/files/16450464/riscv-crypto-spec-vector-extra_v0.0.7.pdf>`__.
```
Comment (dtcxzyw) [inline llvm/docs/RISCVUsage.rst:306]: Can you replace this link with a github repo url+tag?

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 2]
```diff
@@ -150,6 +150,14 @@ let Predicates = [HasStdExtZvkg], RVVConstraint = NoConstraint in {
                  SchedBinaryMC<"WriteVGMULV", "ReadVGMULV", "ReadVGMULV">;
 } // Predicates = [HasStdExtZvkg]
 
+let Predicates = [HasStdExtZvkgs], RVVConstraint = NoConstraint in {
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:153]:
```suggestion
let Predicates = [HasStdExtZvkgs], RVVConstraint = VS2Constraint in {
```
Reserved Encodings
+ SEW is any value other than 32
+ **the vd register group overlaps the vs2 scalar element group**

File: llvm/docs/RISCVUsage.rst (hunk 1 above)
Comment (wangpc-pp) [reply llvm/docs/RISCVUsage.rst:306]: Actually, there is no repo&tag. They are in a pull request.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td (hunk 2 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:153]: Oops, I think we should set this for `vghsh.vv` and `vgmul.vv` as well. cc @4vtomat

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 3]
```diff
@@ -750,8 +750,13 @@ Error RISCVISAInfo::checkDependency() {
     return getError(
         "'zvbc' requires 'v' or 'zve64*' extension to also be specified");
 
-  if ((Exts.count("zvkb") || Exts.count("zvkg") || Exts.count("zvkned") ||
-       Exts.count("zvknha") || Exts.count("zvksed") || Exts.count("zvksh")) &&
+  if (Exts.count("zvbc32e") && !Exts.count("zve32x"))
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:753]: Why is this different than the check for `"'zvk*' requires 'v' or 'zve*' extension to also be specified"`?

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 4]
```diff
@@ -750,8 +750,13 @@ Error RISCVISAInfo::checkDependency() {
...
+    return getError(
+        "'zvbc32e' requires 'v' or 'zve32*' extension to also be specified");
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:755]: Can we say "requires 'v' or 'zve*' extension to also be specified" like the Zvk extension? That way the way the user doesn't need to think about Zve64* implying Zve32x.

File: llvm/docs/RISCVUsage.rst [hunk 5]
```diff
@@ -302,6 +302,9 @@ The primary goal of experimental support is to assist in the process of ratifica
 ``experimental-zicfilp``, ``experimental-zicfiss``
   LLVM implements the `1.0 release specification <https://github.com/riscv/riscv-cfi/releases/tag/v1.0>`__.
 
+``experimental-zvbc32e``, ``experimental-zvkgs``
```
Comment (michaelmaitland) [inline llvm/docs/RISCVUsage.rst:305]: Should we mention these extensions are added in the release notes?

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 6]
```diff
@@ -762,6 +773,14 @@ def HasStdExtZvkg : Predicate<"Subtarget->hasStdExtZvkg()">,
                     AssemblerPredicate<(all_of FeatureStdExtZvkg),
                         "'Zvkg' (Vector GCM instructions for Cryptography)">;
 
+def FeatureStdExtZvkgs
+    : RISCVExperimentalExtension<"zvkgs", 0, 7,
+                                 "'Zvkgs' (Vector-Scalar GCM instructions for Cryptography)",
+                                 [FeatureStdExtZvkg]>;
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:779]: From the Zvkgs speck:

> Zvkgs depends on Zvkg

Does this part of the spec mean that Zvkg must be present too, or does it mean that if Zvkgs is present, then Zvkg is implied?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:779]: The spec is largely written for the hardware design. How a toolchain -march string should work is probably not being thought about. So what the spec is saying is that if your hardware implements Zvkgs it must also implement Zvkg.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td (hunk 2 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:153]: I guess the spec didn't state that for `.vv` version?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:153]: Why? I was thinking it should be the same case?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:153]: I made a comment in https://github.com/riscv/riscv-isa-manual/pull/1306. We may not need this constraint to keep the consistency between `.vv` and `.vs`.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:153]: We already have other instructions that aren't consistent between `.vv` and `.vs`. For example, vaesdf.vv and vaesdf.vs.

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 7]
```diff
@@ -754,6 +754,17 @@ def HasStdExtZvbc : Predicate<"Subtarget->hasStdExtZvbc()">,
                     AssemblerPredicate<(all_of FeatureStdExtZvbc),
                         "'Zvbc' (Vector Carryless Multiplication)">;
 
+def FeatureStdExtZvbc32e
+    : RISCVExperimentalExtension<"zvbc32e", 0, 7,
+                                 "'Zvbc32e' (Vector Carryless Multiplication with 32-bits elements)">;
+def HasStdExtZvbc32e : Predicate<"Subtarget->hasStdExtZvbc32e()">,
```
Comment (4vtomat) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:760]: Does this have any user?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:760]: Currently no, I will remove it.

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Review by topperc — APPROVED** [review]
LGTM

**Review by dtcxzyw — APPROVED** [review]

**Pull request merged**